base/setup/reactos/drivepage.c

```diff
@@ -948,22 +1690,181 @@ DriveDlgProc(
                                     hwndDlg,
                                     MoreOptDlgProc,
                                     (LPARAM)pSetupData);
                     break;
+                }
+
+                case IDC_INITDISK:
+                {
+                    // TODO: Implement disk partitioning initialization
+                    break;
+                }
 
                 case IDC_PARTCREATE:
-                    DialogBoxW(pSetupData->hInstance,
-                               MAKEINTRESOURCEW(IDD_PARTITION),
-                               hwndDlg,
-                               PartitionDlgProc);
+                {
+                    HTLITEM hItem;
+                    PPARTITEM PartItem;
+                    PPARTENTRY PartEntry;
+                    ULONGLONG PartSize;
+                    ULONGLONG MaxPartSize;
+                    ULONG MaxSizeMB;
+                    INT_PTR ret;
+                    PARTCREATE_CTX PartCreateCtx = {0};
+
+                    hList = GetDlgItem(hwndDlg, IDC_PARTITION);
+                    PartItem = GetSelectedPartition(hList, &hItem);
+                    if (!PartItem)
+                    {
+                        /* If the button was clicked, an empty disk
+                         * region should have been selected first */
+                        ASSERT(FALSE);
+                        break;
+                    }
+                    PartEntry = PartItem->PartEntry;
+                    if (PartEntry->IsPartitioned)
+                    {
+                        /* Don't create a partition when one already exists */
+                        ASSERT(FALSE);
+                        break;
+                    }
+                    ASSERT(!PartEntry->Volume);
+
+                    /* Get the partition info stored in the TreeList */
+                    PartCreateCtx.PartItem = PartItem;
+
+                    /* Retrieve the maximum size in MB (rounded up) the partition can have */
+                    MaxPartSize = GetPartEntrySizeInBytes(PartEntry);
+                    MaxSizeMB = (ULONG)RoundingDivide(MaxPartSize, MB);
+                    PartCreateCtx.MaxSizeMB = MaxSizeMB;
+
+                    /* Don't force formatting by default */
+                    PartCreateCtx.ForceFormat = FALSE;
+
+                    /* Show the partitioning dialog */
+                    ret = DialogBoxParamW(pSetupData->hInstance,
+                                          MAKEINTRESOURCEW(IDD_PARTITION),
+                                          hwndDlg,
+                                          PartitionDlgProc,
+                                          (LPARAM)&PartCreateCtx);
+                    if (ret != IDOK)
+                        break;
+
+                    /*
+                     * If the input size, given in MB, specifies the maximum partition
+                     * size, it may slightly under- or over-estimate it due to rounding
```

Comment:
Suggestion:
                     * size, it may be slightly under- or over-estimated due to rounding

Reply:
I think that my confusion was caused by the "it". Is what you mean here ```PartCreateCtx.PartSizeMB```?
Below you do not seem to check for it being greater than MaxSizeMB. It is not obvious why.
So maybe:
* size, PartCreateCtx.PartSizeMB may slightly under-estimate it due to rounding

Reply:
> I think that my confusion was caused by the "it".

I understand, because they refered to two different things.

I'll look at your suggestion of re-expliciting one of the subjects of the sentence.

Reply:
I am fine with you leaving it as it is now. But if you can make it more clear, that would be good.

Reply:
OK my current final suggestion:
> If the input size, given in MB, specifies the maximum partition
size, it may slightly under- or over-estimate the latter [...]

Reply:
OK. That's good for me. Thanks,